### PR TITLE
Fix undefined behavior issues in aarch64

### LINKED ISSUE #74390: ubsan discovers out of bounds array access in runtime/src/native/external/libunwind/src/aarch64/Ginit.c
### Description

Running on linux on ARM64 metal, running the gcperfsim workload, there' a single call into:
```
runtime/src/native/external/libunwind/src/aarch64/Ginit.c:50
50	    return &uc->uc_mcontext.regs[reg];
```
where the value of reg==32, and the array regs is declared to have 31 elements.

However, looking at the containing structure, unw_sigcontext, there's a declaration run like:
          uint64_t regs[31];
        uint64_t sp;
        uint64_t pc;
Which makes a little sense, in that indexing regs[31] will actually pick up the sp, as intended, and then a "shortcut" of regs[32] picks up the pc.  Maybe..

It's 3rd party code.

### Reproduction Steps

Compile for host machine ARM64 with g++-12 -fsanitize  Run gcperfsim.  Observe a runtime error at the source line.

### Expected behavior

no runtime error, perhaps because it is suppressed in the function uc_addr containing the miscreant.

### Actual behavior

a runtime error

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

/cc @AaronRobinsonMSFT

### PR DESCRIPTION
**Pull request by am11**
Fixes #74390.

### CODE REVIEW

File: src/native/external/libunwind/src/aarch64/Ginit.c [hunk 1]
```diff
@@ -46,12 +46,29 @@ unw_addr_space_t unw_local_addr_space = &local_addr_space;
 static inline void *
 uc_addr (unw_tdep_context_t *uc, int reg)
 {
-  if (reg >= UNW_AARCH64_X0 && reg < UNW_AARCH64_V0)
+#ifdef __FreeBSD__
+  if (reg >= UNW_AARCH64_X0 && reg <= UNW_AARCH64_X30)
```
Comment (AaronRobinsonMSFT) [inline src/native/external/libunwind/src/aarch64/Ginit.c:50]: Does upstream need this predicate fix too?

https://github.com/libunwind/libunwind/blob/612ebb80368ef69f713de06eaa3d924403ba9288/src/aarch64/Ginit.c#L50

File: src/native/external/libunwind/src/aarch64/Ginit.c [hunk 2]
```diff
@@ -46,12 +46,29 @@ unw_addr_space_t unw_local_addr_space = &local_addr_space;
...
+  else if (reg == UNW_AARCH64_SP)
+    return &uc->uc_mcontext.sp;
+  else if (reg == UNW_AARCH64_PC)
+    return &uc->uc_mcontext.gp_elr;
+  else if (reg >= UNW_AARCH64_V0 && reg <= UNW_AARCH64_V31)
+    return &GET_FPCTX(uc)->uc_mcontext.mc_fpregs.fp_q[reg - UNW_AARCH64_V0];
+  else
+    return NULL;
+#else /* __FreeBSD__ */
+  if (reg >= UNW_AARCH64_X0 && reg <= UNW_AARCH64_X30)
```
Comment (AaronRobinsonMSFT) [inline src/native/external/libunwind/src/aarch64/Ginit.c:61]: Same here? https://github.com/libunwind/libunwind/blob/612ebb80368ef69f713de06eaa3d924403ba9288/src/aarch64/Ginit.c#L61

File: src/native/external/libunwind/src/aarch64/Ginit.c (hunk 1 above)
Comment (am11) [reply src/native/external/libunwind/src/aarch64/Ginit.c:50]: Yes, I am waiting for CI, we didn't find the issue there, but JIT tests are failing.

**Comment by am11** [comment]
Had to make another change upstream because arm64 checked failed multiple times: https://github.com/libunwind/libunwind/pull/416. I missed LR register in the previous PR.

**Comment by AaronRobinsonMSFT** [comment]
@am11 This looks good to me. Is there any reason to wait on this work?

**Comment by am11** [comment]
(if) It is not urgent, we can wait for upstream PR to merge first. There is no visible difference on linux-arm64, it is just an analyzer fix to make gcc sanitizer happy. We do not go out of bounds of register array (0:31) at callsite.

FreeBSD, OTOH, had a real bug fixed by PR#414 upstream, that lr was accessed by [`gp_x[30]`](https://github.com/freebsd/freebsd-src/blob/373ffc6/sys/arm64/include/ucontext.h#L36-L37) at index 31, but:
* we are currently based on libunwind 1.6.2 snap, that doesn't have the entire `freebsd-arm64` port (which happened in upstream's `master` branch), so no freebsd issues is being fixed by this PR.
* `freebsd-arm64` initial coreclr port is WIP, so nobody is actually blocked on this PR.
* coreclr uses system libunwind on `freebsd` (x64) and not this in-tree copy. Although it is technically possible to use these sources on that platform, it is untested; as it stands, nobody _can be_ unblocked by this PR. :)

**Comment by am11** [comment]
@AaronRobinsonMSFT, upstream PR is merged.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**